File: src/markov_gen.py
import logging
import os
from markovchain.text import MarkovText
from clean import clean_content  # Assuming clean_content is in clean.py

# Ensure the log directory exists
log_directory = 'log'
if not os.path.exists(log_directory):
    os.makedirs(log_directory)

# Set up logging to a file in the log directory
logging.basicConfig(
    filename=os.path.join(log_directory, 'general.log'),
    level=logging.DEBUG,
    format='%(asctime)s - %(levelname)s - %(message)s'
)

def retrieve_posts(client, client_did):
    post_list = []
    has_more = True
    cursor = None  # Initialise cursor for pagination

    logging.info(f"Starting to retrieve posts for client ID: {client_did}")

    while has_more:
        try:
            # Use cursor for pagination
            if cursor:
                data = client.app.bsky.feed.post.list(client_did, limit=100, cursor=cursor)
            else:
                data = client.app.bsky.feed.post.list(client_did, limit=100)

            logging.debug(f"Fetched data with cursor: {cursor}")

            # Log the structure of the returned data
            logging.debug("Data response structure: %s", data)

            # Check for the correct attribute containing the posts
            if not hasattr(data, 'records') or not data.records:
                logging.info("No more posts found or 'records' attribute is missing.")
                break

            # Fetch posts from the 'records' attribute
            posts = data.records

            # Add the posts to the post_list
            post_list.extend(posts.values())  # Access values directly from the dictionary

            logging.info(f"Retrieved {len(posts)} posts.")

            # Get the next cursor for pagination
            cursor = data.cursor  # Use the 'cursor' returned from the data for the next page
            has_more = bool(cursor)  # Continue if there is a next_cursor

        except Exception as e:
            logging.error(f"Error fetching posts: {e}")
            break

    logging.info(f"Completed retrieval of posts for client ID: {client_did}. Total posts retrieved: {len(post_list)}")

    # Log the structure of the first few posts for debugging
    for post in post_list[:5]:  # Show only the first 5 posts for brevity
        logging.debug("Post structure: %s", post)

    return post_list

def refresh_dataset(markov, source_posts):
    if not markov:
        markov = MarkovText()  # Initialise if not passed as an argument

    if source_posts:
        logging.info(f"Fetched {len(source_posts)} original posts and replies from the source account.")

    for post in source_posts:
        if isinstance(post, str):
            markov.data(post, part=True)  # Directly add the string if it's a simple string
        elif hasattr(post, 'text'):
            markov.data(post.text, part=True)  # Add if it's an object with text
        elif hasattr(post, 'value') and hasattr(post.value, 'text'):
            markov.data(post.value.text, part=True)  # Access text correctly
        else:
            logging.warning("Post does not have 'text' attribute, skipping: %s", post)

    markov.data('', part=False)  # Ensuring a proper closure of data input

    # Log the current dataset state for debugging
    logging.debug("Current length of posts: %s\n\n\n", len(source_posts))
    logging.debug("Current Markov dataset: %s", markov.storage)

    return markov

def generate(markov, char_limit):
    try:
        generated_text = markov()
    except KeyError as e:
        logging.error("KeyError occurred during generation: %s", e)
        return []

    if len(generated_text) > char_limit:
        generated_text = generated_text[:char_limit]

    logging.debug("Generated Text: %s", generated_text)

    words = generated_text.split()
    return words

def get_account_posts(client, client_did):
    posts = retrieve_posts(client, client_did)

    # Debugging: Print structure of the first post
    if posts:
        logging.debug("First post structure: %s", posts[0])

    # Ensure we are accessing the text correctly
    cleaned_posts = []
    for post in posts:
        if hasattr(post, 'value') and hasattr(post.value, 'text'):
            cleaned_text = clean_content(post.value.text)
            cleaned_posts.append(cleaned_text)
        elif hasattr(post, 'text'):  # Fall back to checking for a direct 'text' attribute
            cleaned_text = clean_content(post.text)
            cleaned_posts.append(cleaned_text)
        else:
            logging.warning("Post does not have 'value' or 'text', skipping: %s", post)

    return cleaned_posts

File: src/markov_gen.rs
//! Fetch source posts, build a Markov chain, and generate text.
//!
//! The pipeline: paginate all posts from the source DID → clean each one
//! → feed into a `markov::Chain` → generate a string capped at `char_limit`.

use bsky_sdk::BskyAgent;
use crate::clean::clean_content;
use atrium_api::types::string::Did;
use markov::Chain;
use anyhow::{Result, anyhow};
use tracing::{info, debug};

/// Retrieve all posts from the source account, cleaned and ready for training.
///
/// Paginates through the AT Protocol `list_records` endpoint (100 records
/// per page). Only `app.bsky.feed.post` records are fetched — other
/// collection types are ignored.
pub async fn retrieve_posts(agent: &BskyAgent, did: Did) -> Result<Vec<String>> {
    let mut post_list = Vec::new();
    let mut cursor = None;

    info!("Starting to retrieve posts for client ID: {}", did.as_str());

    loop {
        let output = agent.api.com.atproto.repo.list_records(
            atrium_api::com::atproto::repo::list_records::ParametersData {
                collection: "app.bsky.feed.post".parse().map_err(|e| anyhow!("{:?}", e))?,
                repo: did.as_str().parse().map_err(|e| anyhow!("{:?}", e))?,
                cursor: cursor.clone(),
                limit: Some(100.try_into().map_err(|e| anyhow!("{:?}", e))?),
                reverse: None,
            }.into()
        ).await?;

        let data = output.data;
        let records_len = data.records.len();
        for record in &data.records {
            if let Ok(post_record) = serde_json::from_value::<atrium_api::app::bsky::feed::post::Record>(serde_json::to_value(&record.value)?) {
                post_list.push(clean_content(&post_record.text));
            }
        }

        info!("Retrieved {} posts.", records_len);

        cursor = data.cursor;
        if cursor.is_none() {
            break;
        }
    }

    info!("Completed retrieval of posts for client ID: {}. Total posts retrieved: {}", did.as_str(), post_list.len());
    Ok(post_list)
}

/// Feed cleaned posts into a fresh Markov chain, replacing the old dataset.
///
/// Creates a new `Chain` each call rather than incrementally updating —
/// the corpus is small enough that rebuilding is simpler than tracking
/// individual record additions and deletions.
pub fn refresh_dataset(source_posts: Vec<String>) -> Chain<String> {
    let mut chain = Chain::new();
    for post in source_posts {
        chain.feed_str(&post);
    }
    info!("Markov dataset refreshed.");
    chain
}

/// Generate a single post from the chain, truncated to `char_limit`.
///
/// Returns an empty string if the chain has no data yet (first-run edge case).
pub fn generate_text(chain: &Chain<String>, char_limit: usize) -> String {
    let generated = chain.generate_str();
    if generated.is_empty() {
        return String::new();
    }

    let mut result = generated;
    if result.len() > char_limit {
        result.truncate(char_limit);
    }
    debug!("Generated Text: {}", result);
    result
}
